File: app/src/androidTest/java/com/example/win/easy/view/fragment/SongListFragmentTest.java
package com.example.win.easy.view.fragment;

import android.os.Bundle;

import androidx.fragment.app.FragmentFactory;
import androidx.fragment.app.testing.FragmentScenario;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner;

import com.example.win.easy.BooleanSemaphore;
import com.example.win.easy.R;
import com.example.win.easy.display.DisplayServiceAdapter;
import com.example.win.easy.download.DownloadServiceAdapter;
import com.example.win.easy.value_object.SongListVO;
import com.example.win.easy.value_object.SongVO;
import com.example.win.easy.viewmodel.SongListViewModel;
import com.example.win.easy.web.callback.OnReadyFunc;
import com.qmuiteam.qmui.alpha.QMUIAlphaImageButton;
import com.qmuiteam.qmui.widget.QMUITopBar;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.CoreMatchers.allOf;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4ClassRunner.class)
public class SongListFragmentTest {


    @InjectMocks SongListFragment songListFragment;
    @Mock DisplayServiceAdapter displayServiceAdapter;
    @Mock DownloadServiceAdapter downloadServiceAdapter;
    @Mock FragmentFactory fragmentFactory;
    @Mock ViewModelProvider.Factory viewModelFactory;
    @Mock SongListViewModel songListViewModel;
    @Mock NavController mockNavController;


    /**
     * <p>简陋地测试下界面是不是正常显示</p>
     * <p>把带字的、特定类型的元素确认下isDisplayed就完了</p>
     */
    @Test
    public void testDisplay(){
        onView(allOf(isDescendantOfA(instanceOf(QMUITopBar.class)),withText(selectedSongList.getName()))).check(matches(isDisplayed()));
        onView(instanceOf(QMUIAlphaImageButton.class)).check(matches(isDisplayed()));
        onView(withText(downloadedSong.getName())).check(matches(isDisplayed()));
        onView(withText(nonDownloadedSongWithUrl.getName())).check(matches(isDisplayed()));
    }

    /**
     * <p>测试当点击一首已经下载好的歌曲时，直接播放</p>
     */
    @Test
    public void testClickDownloadedSong(){
        onView(withText(downloadedSong.getName())).perform(click());
        verify(displayServiceAdapter,times(1)).startWith(downloadedSong,songsInSelectedSongListLiveData.getValue());
    }


    /**
     * <p>测试当点击一首还没下载的歌曲时，先触发下载，然后等下载结束后，触发播放</p>
     */
    @Test
    public void testClickDownloadedSongWith(){
        onView(withText(nonDownloadedSongWithUrl.getName())).perform(click());
        verify(downloadServiceAdapter,times(1)).download(eq(nonDownloadedSongWithUrl),any(OnReadyFunc.class));
        verify(displayServiceAdapter,times(1)).startWith(downloadedSongFromUrl,songsInSelectedSongListLiveData.getValue());
    }

    /**
     * <p>测试当点击右上角的按钮时，触发到“添加歌曲到歌单”fragment的导航</p>
     */
    @Test
    public void testAddSongToSongList(){
        onView(instanceOf(QMUIAlphaImageButton.class)).perform(click());
        verify(mockNavController,times(1)).navigate(SongListFragmentDirections.actionSongListFragmentToAddSongToSongListFragment(selectedSongList));
    }

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        setUpData();

        mockDownloadService();
        mockLiveDataFromViewModel();
        mockViewModelFactory();
        mockFragmentFactory();

        launchTestedFragment();
    }

    private void setUpData(){
        setUpLiveData();
        setUpBundleData();
        setUpDownloadedData();
    }

    private void setUpLiveData(){
        List<SongVO> songsInSelectedSongList=new ArrayList<>();
        songsInSelectedSongList.add(downloadedSong);
        songsInSelectedSongList.add(nonDownloadedSongWithUrl);
        songsInSelectedSongListLiveData.postValue(songsInSelectedSongList);
    }

    private void setUpBundleData(){
        bundlePassedFromSelection.putSerializable("selectedSongList", selectedSongList);
    }

    private void setUpDownloadedData(){
        downloadedSongFromUrl=SongVO.builder()
                .name(nonDownloadedSongWithUrl.getName())
                .songFileUrl(nonDownloadedSongWithUrl.getSongFileUrl())
                .avatarPath(nonDownloadedSongWithUrl.getAvatarPath())
                .songFilePath("/data/user/0/downloaded.mp3")
                .build();
    }

    private void mockDownloadService() {
        doAnswer(invocation -> {

            //获取参数
            SongVO songToDownload=invocation.getArgument(0);
            OnReadyFunc<SongVO> onReadyFunc=invocation.getArgument(1);

            //模拟发起请求
            Executors.newSingleThreadExecutor().execute(this::mockDownloadTask);

            waitForDownloadFinished();
            onReadyFunc.onReady(downloadedSongFromUrl);

            return null;
        }).when(downloadServiceAdapter).download(any(SongVO.class),any(OnReadyFunc.class));
    }

    private void mockDownloadTask(){
        downloading();
        notifyDownloadFinished();
    }

    private void downloading(){
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void mockLiveDataFromViewModel() {
        when(songListViewModel.songsOf(selectedSongList)).thenReturn(songsInSelectedSongListLiveData);
    }

    private void mockViewModelFactory() {
        when(viewModelFactory.create(SongListViewModel.class)).thenReturn(songListViewModel);
    }

    private void mockFragmentFactory() {
        when(fragmentFactory.instantiate(any(ClassLoader.class),eq(SongListFragment.class.getName()))).thenReturn(songListFragment);
    }

    private void launchTestedFragment(){
        FragmentScenario<SongListFragment> scenario = FragmentScenario.launchInContainer(SongListFragment.class, bundlePassedFromSelection, R.style.AppTheme, fragmentFactory);
        scenario.onFragment(fragment -> Navigation.setViewNavController(fragment.requireView(), mockNavController));
    }

    private void waitForDownloadFinished(){
        synchronized (downloadFinished) {
            synchronized (mainThreadWaiting) {
                mainThreadWaiting.setReady(true);
            }
            try {
                downloadFinished.wait();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    private void notifyDownloadFinished(){
        while (true){
            synchronized (mainThreadWaiting){
                if (mainThreadWaiting.isReady())
                    break;
            }
        }
        synchronized (downloadFinished){
            downloadFinished.notifyAll();
        }
    }

    //LiveData和这个界面的数据
    private SongListVO selectedSongList=SongListVO.builder().name("体测").build();
    private SongVO downloadedSong=SongVO.builder().name("电灯胆 - 邓丽欣").songFilePath("/songs/d.mp3").build();
    private SongVO nonDownloadedSongWithUrl=SongVO.builder().name("藏在贴纸相背后 - 陈奕迅").songFileUrl("https://www.google.com/songs/song.mp3").build();
    private MutableLiveData<List<SongVO>> songsInSelectedSongListLiveData=new MutableLiveData<>();


    //Bundle数据
    private Bundle bundlePassedFromSelection=new Bundle();

    //下载的数据
    private SongVO downloadedSongFromUrl;

    //用于同步的数据
    private final BooleanSemaphore mainThreadWaiting =new BooleanSemaphore(false);
    private final Object downloadFinished=new Object();
}

File: app/src/androidTest/java/com/example/win/easy/view/main/SongListFragmentTest.kt
package com.example.win.easy.view.main

import android.os.Bundle
import android.util.Log
import android.widget.ProgressBar
import androidx.fragment.app.FragmentFactory
import androidx.fragment.app.testing.FragmentScenario
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import com.example.win.easy.BooleanSemaphore
import com.example.win.easy.R
import com.example.win.easy.TestCoroutineRule
import com.example.win.easy.display.interfaces.DisplayService
import com.example.win.easy.download.DownloadServiceAdapter
import com.example.win.easy.repository.db.data_object.SongDO
import com.example.win.easy.repository.db.data_object.SongListDO
import com.example.win.easy.viewmodel.SongListViewModelImpl
import com.qmuiteam.qmui.alpha.QMUIAlphaImageButton
import com.qmuiteam.qmui.widget.QMUITopBar
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.impl.annotations.RelaxedMockK
import kotlinx.coroutines.*
import kotlinx.coroutines.test.TestCoroutineDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runBlockingTest
import kotlinx.coroutines.test.setMain
import org.hamcrest.CoreMatchers
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.coroutines.CoroutineContext

@ExperimentalCoroutinesApi
class SongListFragmentTest {
    @InjectMockKs lateinit var songListFragment: SongListFragment
    @RelaxedMockK lateinit var displayService: DisplayService
    @MockK lateinit var downloadServiceAdapter: DownloadServiceAdapter
    @MockK lateinit var fragmentFactory: FragmentFactory
    @MockK lateinit var viewModelFactory: ViewModelProvider.Factory
    private val songListViewModel= mockkClass(SongListViewModelImpl::class,relaxed = true)
    @MockK(relaxUnitFun = true) lateinit var mockNavController: NavController

    @get:Rule
    val testRule=TestCoroutineRule()

    /**
     * 测试当点击右上角的按钮时，触发到“添加歌曲到歌单”fragment的导航
     */
    @Test
    fun testAddSongToSongList() {
        scenario.onFragment { fragment: SongListFragment -> Navigation.setViewNavController(fragment.requireView(), mockNavController) }
        Espresso.onView(CoreMatchers.instanceOf(QMUIAlphaImageButton::class.java)).perform(ViewActions.click())
        verify{ mockNavController.navigate(SongListFragmentDirections.actionSongListFragmentToAddSongToSongListFragment(selectedSongList)) }
    }

    private val mSpinner=MutableLiveData(false)
    private val mSnackbar=MutableLiveData("")
    private val onlineSongsLiveData=MutableLiveData<List<SongDO>>()

    /**
     * 测试界面启动
     * 1. 测试界面元素、本地歌单的正常显示
     * 2. 网络歌单测试：
     *      1. 加载时的状态
     *      2. 加载成功时的界面
     */
    @Test
    fun testFragmentLaunch(): Unit =runBlocking{

        // 本地歌单以及界面元素测试
        every { songListViewModel.loadSongsIn(selectedSongList) } returns songsInSelectedSongList
        scenario.recreate()
        Espresso.onView(CoreMatchers.allOf(ViewMatchers.isDescendantOfA(CoreMatchers.instanceOf(QMUITopBar::class.java)), ViewMatchers.withText(selectedSongList.name))).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        Espresso.onView(CoreMatchers.instanceOf(QMUIAlphaImageButton::class.java)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
        checkItemWithName(downloadedSong.name)
        checkItemWithName(nonDownloadedSongWithUrl.name)

        // 网络歌单测试

        // 加载中
        every { songListViewModel.loadSongsIn(selectedSongList) }.answers{
            mSpinner.postValue(false)
            onlineSongsLiveData
        }
        scenario.recreate()
        Espresso.onView(CoreMatchers.instanceOf(ProgressBar::class.java)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))

        // 加载完
        mSpinner.postValue(true)
        onlineSongsLiveData.postValue(listOf(downloadedSong,nonDownloadedSongWithUrl))
        checkItemWithName(downloadedSong.name)
        checkItemWithName(nonDownloadedSongWithUrl.name)
    }

    private fun checkItemWithName(name: String){
        Espresso.onView(ViewMatchers.withText(name)).check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
    }

    private var blockCap = CapturingSlot<()->Unit>()
    private var songCap = CapturingSlot<SongDO>()

    /**
     * 测试点击下载，且下载后触发播放
     */
    @Test
    fun testClickSongToDownload()= testRule.testCoroutineDispatcher.runBlockingTest{
        every { downloadServiceAdapter.download(capture(songCap),capture(blockCap)) } answers {
            launch(testRule.testCoroutineDispatcher) {
                yield()
                songCap.captured.songPath="Test Path"
                withContext(Dispatchers.Main){
                    blockCap.captured()
                }
            }
        }
        every { songListViewModel.loadSongsIn(selectedSongList) } returns songsInSelectedSongList
        scenario.recreate()
        Espresso.onView(ViewMatchers.withText(nonDownloadedSongWithUrl.name)).perform(ViewActions.click())
        assertEquals(null,nonDownloadedSongWithUrl.songPath)
        verify(exactly = 0) { displayService.restartWith(nonDownloadedSongWithUrl,selectedSongList,songsInSelectedSongList.value!!) }

        runCurrent()

        assertEquals("Test Path",nonDownloadedSongWithUrl.songPath)
        verify(exactly = 1) { displayService.restartWith(nonDownloadedSongWithUrl,selectedSongList,songsInSelectedSongList.value!!) }

        nonDownloadedSongWithUrl.songPath=null
    }

    @Before
    fun setUp() {
        every { songListViewModel.spinner } returns mSpinner
        every { songListViewModel.snackbar } returns mSnackbar

        MockKAnnotations.init(this)

        every { downloadServiceAdapter.spinner } returns mSpinner
        every { downloadServiceAdapter.snackbar } returns mSnackbar

        every { viewModelFactory.create(SongListViewModelImpl::class.java) } returns songListViewModel
        every { fragmentFactory.instantiate(any(),SongListFragment::class.java.name) } returns songListFragment
        scenario = FragmentScenario.launchInContainer(SongListFragment::class.java, bundlePassedFromSelection, R.style.AppTheme, fragmentFactory)
    }


    //LiveData和这个界面的数据
    private val selectedSongList = SongListDO().apply {
        name="体测"
    }
    private val downloadedSong = SongDO().apply{
        name="电灯胆 - 邓丽欣"
        songPath="/songs/d.mp3"
    }
    private val nonDownloadedSongWithUrl = SongDO().apply{
        name="藏在贴纸相背后 - 陈奕迅"
        songUrl="https://www.google.com/songs/song.mp3"
    }

    //下载的数据
    private var downloadedSongFromUrl=SongDO().apply {
        name=nonDownloadedSongWithUrl.name
        songUrl=nonDownloadedSongWithUrl.songUrl
        avatarPath=nonDownloadedSongWithUrl.avatarPath
        songPath="/data/user/0/downloaded.mp3"
    }

    //Bundle数据
    private val bundlePassedFromSelection = Bundle().apply { putSerializable("selectedSongList", selectedSongList) }

    private val songsInSelectedSongList=MutableLiveData(listOf(downloadedSong,nonDownloadedSongWithUrl))
    private lateinit var scenario: FragmentScenario<SongListFragment>
}